### ISSUE TITLE
`torch._numpy.random.shuffle` can duplicate or drop rows for `torch.Tensor` with NumPy random stream

### ISSUE THREAD

**Issue by Ei9en** [pytorch/pytorch#197795]
### 🐛 Describe the bug

# `torch._numpy.random.shuffle` can duplicate or drop rows for `torch.Tensor` with NumPy random stream

### 🐛 Describe the bug

`torch._numpy.random.shuffle` can produce duplicates and drop elements when:

- the input is a `torch.Tensor`, and
- `torch._dynamo.config.use_numpy_random_stream=True`.

Instead of only permuting the tensor along its first axis, the contents can be corrupted.

The call also emits a NumPy warning indicating that shuffling tensor-like objects with view semantics is not guaranteed to behave correctly.

### Reproduction

```python
import torch
import torch._numpy as tnp
import torch._dynamo.config as config

x = torch.arange(12).reshape(6, 2)
expected = sorted(map(tuple, x.tolist()))

print("before:")
print(x)

with config.patch(use_numpy_random_stream=True):
    tnp.random.seed(0)
    tnp.random.shuffle(x)

print("\nafter:")
print(x)

print(
    "\nrows preserved:",
    sorted(map(tuple, x.tolist())) == expected,
)
```

On my machine this produces:

```text
before:
tensor([[ 0,  1],
        [ 2,  3],
        [ 4,  5],
        [ 6,  7],
        [ 8,  9],
        [10, 11]])

UserWarning: you are shuffling a 'Tensor' object which is not a subclass
of 'Sequence'; `shuffle` is not guaranteed to behave correctly. E.g.,
non-numpy array/tensor objects with view semantics may contain duplicates
after shuffling.

after:
tensor([[0, 1],
        [2, 3],
        [2, 3],
        [6, 7],
        [0, 1],
        [8, 9]])

rows preserved: False
```

The resulting tensor is not a permutation of the original input: some rows
are duplicated while others disappear.

### Expected behavior

`torch._numpy.random.shuffle` should only reorder elements along the first axis.

In particular, the multiset of rows should be preserved:

```python
sorted(map(tuple, x_after.tolist())) == sorted(
    map(tuple, x_before.tolist())
)
```

should be `True`.

### Additional context

`torch._numpy.random.shuffle` explicitly accepts both `torch.Tensor` and
`torch._numpy.ndarray`.

When `use_numpy_random_stream=False`, the implementation uses a PyTorch
permutation and copies the result back into the tensor.

When `use_numpy_random_stream=True`, the `deco_stream` wrapper dispatches to
`numpy.random.shuffle`. `torch._numpy.ndarray` arguments are converted to
NumPy arrays by the wrapper, while a raw `torch.Tensor` is passed through
unchanged.

There also appears to be a coverage gap in `test/torch_np/test_random.py`:
the `TestShuffle` tests are parameterized with `use_numpy=[True, False]`,
but the parameter is currently not applied through
`control_stream(use_numpy)`, so both variants appear to exercise the
default PyTorch random stream.

### Environment

Reproduced with:

- PyTorch 2.2.2
- Python 3.12
- macOS

The corresponding `torch._numpy.random` dispatch path is still present in
the current source tree.

### Versions

I can prepare a small regression test and fix for this once the issue is marked actionable.

cc @mruberry @rgommers @pbelevich

**Comment by malfet** [comment]
`torch._numpy` is a private API, used primarily to help inductor reason about those ops. 
Plus it already prints a warning, that it might result in duplicates.

**Comment by Ei9en** [comment]
Gotcha, so this is more of a conceptual consistency issue than a practical one, and probably not worth fixing then. Thanks for clarifying.
